src/Opaleye/TH.hs

```diff
@@ -0,0 +1,652 @@
+            [primaryField] -> Just $ makeFieldName modelName (columnName primaryField)
+            _ -> Nothing
+  
+makeTablename :: String -> String
+makeTablename t = t ++ "Table"
+
+makePGReadTypeName :: String -> String
+makePGReadTypeName tn = tn ++ "PGRead"
+
+makePGWriteTypeName :: String -> String
+makePGWriteTypeName tn = tn ++ "PGWrite"
+
+makeAdapterName :: String -> String
+makeAdapterName tn = 'p':tn
+
+makeOpaleyeModels :: (ConnectInfo, Options) -> Q [Dec]
+makeOpaleyeModels (ci, op) = makeOpaleyeModels' (ci, op, [])
+
+makeOpaleyeModels' :: Env -> Q [Dec]
+makeOpaleyeModels' env = fst <$> runStateT (do
+  newTypeDecs <- makeNewTypes
+  (_, options, _) <- get
+  let names = fst <$> tableOptions options
+  let models = (modelName.snd) <$> tableOptions options
+  instances <- makeNewtypeInstances
+  decs <- concat <$> zipWithM makeOpaleyeModel names models
+  return $ newTypeDecs ++ decs ++ instances) env
+
+collectNewTypes :: EnvM [(ColumnInfo, String)]
+collectNewTypes = do
+  (connInfo, options, _) <- get
+  concat <$> mapM getNewTypes (tableOptions options)
+  where
+    getNewTypes :: (String, TableOptions) -> EnvM [(ColumnInfo, String)]
+    getNewTypes (tbName, tbOptions) = do
+      (connectInfo, _, _) <- get
+      fieldInfos <- (lift.runIO) $ do
+        conn <- connect connectInfo
+        getColumns conn tbName
+      return $ fromJust <$> (filter isJust $ (tryNewType tbOptions) <$> fieldInfos)
+    tryNewType :: TableOptions -> ColumnInfo -> Maybe (ColumnInfo, String)
+    tryNewType to ci = (\n -> (ci, n)) <$> lookup (columnName ci) (overrideDefaultTypes to)
+
+makeNewTypes :: EnvM [Dec]
+makeNewTypes = do
+  nts <- collectNewTypes
+  (a, b) <- foldM makeNewType ([], []) nts
+  return b
+  where
+  makeNewType :: ([String], [Dec]) -> (ColumnInfo, String) -> EnvM ([String], [Dec])
+  makeNewType (added, decs) (ci, nt_name) = do
+    if nt_name `elem` added then (return (added, decs)) else do
+      dec <- makeNewType' nt_name
+      return (nt_name:added, dec:decs)
+    where
+      makeNewType' :: String -> EnvM Dec
+      makeNewType' name = do
+        let bang = Bang NoSourceUnpackedness NoSourceStrictness
+        haskellType <- makeRawHaskellType ci
+        return $ NewtypeD [] (mkName name) [] Nothing (NormalC (mkName name) [(bang, haskellType)]) [ConT ''Show]
+
+makeFieldName :: String -> String -> String
+makeFieldName modelName (s:ss) = "_" ++ (toLower <$> modelName) ++ (toUpper s:replaceUnderscore ss)
+
+replaceUnderscore ('_':x:xs) = (toUpper x):(replaceUnderscore xs)
+replaceUnderscore ('_':xs) = (replaceUnderscore xs)
+replaceUnderscore (x:xs) = x:(replaceUnderscore xs)
+replaceUnderscore [] = ""
+
+ucFirst :: String -> String
+ucFirst (s:ss) = (toUpper s):ss
+    
+makeOpaleyeModel :: String -> String -> EnvM [Dec]
+makeOpaleyeModel t r = do
+  (connectInfo, _, _) <- get
+  let recordName = mkName r
+  let recordPolyName = mkName $ r ++ "Poly"
+  fieldInfos <- (lift.runIO) $ do
```

Comment:
Querying for column info? Have changed it so that it only has to be done once for each top level TH call.